runtime/io/common/src/aws/smithy/kotlin/runtime/io/SdkByteBuffer.kt

```diff
@@ -0,0 +1,452 @@
+internal expect fun Memory.Companion.ofByteArray(src: ByteArray, offset: Int = 0, length: Int = src.size - offset): Memory
+
+/**
+ * A buffer with read and write positions. Similar in spirit to `java.nio.ByteBuffer` but for use
+ * in Kotlin Multiplatform.
+ *
+ * Unlike `ByteBuffer`, this buffer will implicitly grow as needed to fulfill write requests.
+ * However, explicitly reserving the required space up-front before a series of writes will be
+ * more efficient.
+ *
+ * **concurrent unsafe**: Do not read/write using the same [SdkByteBuffer] instance from different threads.
+ */
+@OptIn(ExperimentalIoApi::class)
+@InternalApi
+class SdkByteBuffer internal constructor(
+    // we make use of ktor-io's `Memory` type which already implements most of the functionality in a platform
+    // agnostic way. We just need to wrap some methods around it
+    internal var memory: Memory,
+    val isReadOnly: Boolean = false,
+    val allowReallocation: Boolean = true
+) : Buffer, MutableBuffer {
+    constructor(initialCapacity: ULong, readOnly: Boolean = false, allowReallocation: Boolean = true) : this(DefaultAllocator.alloc(initialCapacity), readOnly, allowReallocation)
+
+    companion object {
+        /**
+         * Create an SdkBuffer backed by the given ByteArray.
+         *
+         * NOTE: Care should be taken if using this as a writable buffer to not exceed the capacity. Doing so will
+         * cause a [FixedBufferSizeException] to be thrown.
+         *
+         * @param src the ByteArray to wrap
+         * @param offset the offset into the array that will mark the first byte written/read
+         * @param length the number of bytes to make available for reading/writing
+         */
+        fun of(
+            src: ByteArray,
+            offset: Int = 0,
+            length: Int = src.size - offset,
+        ): SdkByteBuffer = SdkByteBuffer(Memory.ofByteArray(src, offset, length), allowReallocation = false)
+    }
+
+    private val state = SdkBufferState()
+
+    /**
+     * The total capacity of the buffer
+     */
+    val capacity: ULong
+        get() = memory.size.toULong()
+
+    /**
+     * The current read position. Always non-negative and <= [writePosition].
+     */
+    val readPosition: ULong
+        get() = state.readHead
+
+    /**
+     * The current write position. Will never run ahead of [capacity] and always greater than [readPosition]
+     */
+    val writePosition: ULong
+        get() = state.writeHead
+
+    /**
+     * Number of bytes available for reading
+     */
+    override val readRemaining: ULong
+        get() = writePosition - readPosition
+
+    /**
+     * Free space left for writing
+     */
+    override val writeRemaining: ULong
+        get() = capacity - writePosition
+
+    /**
+     * Reserve capacity for at least [count] bytes to be written.
+     *
+     * More than [count] bytes may be allocated in order to avoid frequent re-allocations.
+     */
+    fun reserve(count: Long) {
```

Comment:
**Question**: Why signed type for `count`?

Reply:
There's not good choice here actually. As I mentioned in the other comment the real constraints for the underlying memory are usually `Int`. Both `Long` and `ULong` could be unsatisfiable when exceeding `Int.MAX_VALUE`.

I don't remember the details on why I changed this to `Long` now that I'm looking at it. I'm open to ideas but want to hear how we want to handle unsatisfiable requests. My inclination is almost to return this to `Int`.

Reply:
@ianbotsf Same, any opinion or final input?